### PR TITLE
Add groupBy for NonEmptySet

### PR DESCRIPTION
**Pull request by Obarros**
Should fixe #2282.

@LukaJCB PTAL.

### CODE REVIEW

File: core/src/main/scala/cats/data/NonEmptySet.scala [hunk 1]
```diff
@@ -346,6 +346,21 @@ sealed class NonEmptySetOps[A](val value: NonEmptySet[A]) {
     */
   def zipWithIndex: NonEmptySet[(A, Int)] =
     NonEmptySetImpl.create(toSortedSet.zipWithIndex)
+
+  /**
+    * Groups elements inside this `NonEmptySet` according to the `Order`
+    * of the keys produced by the given mapping function.
+    */
+  def groupBy[B](f: A => B)(implicit B: Order[B]): NonEmptyMap[B, NonEmptySet[A]] = {
```
Comment (kubukoz) [inline core/src/main/scala/cats/data/NonEmptySet.scala]: If we're returning `NonEmptyMap` here, does it make sense to make `NonEmptyList#groupBy` also return a non-empty map? @LukaJCB

Comment (LukaJCB) [reply core/src/main/scala/cats/data/NonEmptySet.scala]: Yeah that's probably true, but we can't change it without breaking compatibility

**Comment by kubukoz** [comment]
Right. I'm sooo excited for the 2.x branch ;)

**Comment by kailuowang** [comment]
to make this available with `import cats.implicits._` , we also need to add this to this
[binary compatible trait](https://github.com/typelevel/cats/blob/d4bc3493d51c1a42c29e5eaed7922729f24286ad/core/src/main/scala/cats/syntax/all.scala#L62)

**Comment by LukaJCB** [comment]
@kailuowang that's already there, no? https://github.com/typelevel/cats/pull/2285/files#diff-97bffe9b5ca6c2e744d09d9ce308043dR67

**Comment by kailuowang** [comment]
Sorry I missed that.

**Comment by LukaJCB** [comment]
What we did forget is to add a specific import for these syntax extensions, we should add an `object set extends SetSyntax` to here: https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/syntax/package.scala#L46

@Obarros

**Comment by Obarros** [comment]
@LukaJCB any test that we can have for this?

**Comment by kubukoz** [comment]
@Obarros I think there is a spec that checks if syntax compiles. Look for `SyntaxSuite` :)

**Comment by Obarros** [comment]
Ping @kubukoz @LukaJCB!
Updated the `SyntaxSuite` with list and set syntax.

**Comment by kubukoz** [comment]
looks good, :shipit:

File: tests/src/test/scala/cats/tests/SyntaxSuite.scala [hunk 2]
```diff
@@ -358,5 +360,22 @@ object SyntaxSuite extends AllSyntaxBinCompat with AllInstances with AllSyntax {
...
+
+  def testNonEmptySet[A: Order, B: Order] : Unit = {
+    val f = mock[A => B]
+    val set = mock[SortedSet[A]]
+
+    val nes: Option[NonEmptySet[A]] = set.toNes
+    val grouped: SortedMap[B, NonEmptySet[A]] = set.groupByNes(f)
+  }
+
+  def testNonEmptyList[A: Order, B: Order] : Unit = {
```
Comment (LukaJCB) [inline tests/src/test/scala/cats/tests/SyntaxSuite.scala]: I think we don't need this `Order` constraint on `A`, leaving it on `B` should be enough. :)

Comment (Obarros) [reply tests/src/test/scala/cats/tests/SyntaxSuite.scala]: Thanks, @LukaJCB.
Let me remove it, I will also rebase on top of the latest master before squashing the commits and push.

**Comment by kailuowang** [comment]
@obarros, it's no big deal but it's not necessary to squash before merge. We usually do squash merge (meaning all commits on pr will be squashed on merge by GitHub) . Having separate commits helps reviewing the latest changes.

**Pull request merged**